pkg/dataobj/builder.go

```diff
@@ -286,75 +249,47 @@ func streamSizeEstimate(stream logproto.Stream) int {
 	return size
 }
 
-// Flush flushes all buffered data to object storage. Calling Flush can result
+// Flush flushes all buffered data to the buffer provided. Calling Flush can result
 // in a no-op if there is no buffered data to flush.
 //
-// If Flush builds an object but fails to upload it to object storage, the
-// built object is cached and can be retried. [Builder.Reset] can be called to
-// discard any pending data and allow new data to be appended.
-func (b *Builder) Flush(ctx context.Context) (FlushResult, error) {
-	buf, err := b.FlushToBuffer()
+// [Builder.Reset] is called after a successful Flush to discard any pending data and allow new data to be appended.
+func (b *Builder) Flush(output *bytes.Buffer) (FlushStats, error) {
+	err := b.buildObject(output)
 	if err != nil {
-		return FlushResult{}, fmt.Errorf("flushing buffer: %w", err)
+		return FlushStats{}, fmt.Errorf("building object: %w", err)
 	}
 
-	timer := prometheus.NewTimer(b.metrics.flushTime)
-	defer timer.ObserveDuration()
-
-	sum := sha256.Sum224(b.flushBuffer.Bytes())
-	sumStr := hex.EncodeToString(sum[:])
-
-	objectPath := fmt.Sprintf("tenant-%s/objects/%s/%s", b.tenantID, sumStr[:b.cfg.SHAPrefixSize], sumStr[b.cfg.SHAPrefixSize:])
-	if err := b.bucket.Upload(ctx, objectPath, bytes.NewReader(buf.Bytes())); err != nil {
-		return FlushResult{}, fmt.Errorf("uploading object: %w", err)
-	}
-
-	minTime, maxTime := b.streams.GetBounds()
+	minTime, maxTime := b.streams.TimeRange()
 
 	b.Reset()
-	return FlushResult{
-		Path:         objectPath,
+	return FlushStats{
 		MinTimestamp: minTime,
 		MaxTimestamp: maxTime,
 	}, nil
 }
 
-func (b *Builder) FlushToBuffer() (*bytes.Buffer, error) {
-	switch b.state {
-	case builderStateEmpty:
-		return nil, nil // Nothing to flush
-	case builderStateDirty:
-		if err := b.buildObject(); err != nil {
-			return nil, fmt.Errorf("building object: %w", err)
-		}
-		b.state = builderStateFlush
-	}
-
-	return b.flushBuffer, nil
-}
-
-func (b *Builder) buildObject() error {
+func (b *Builder) buildObject(output *bytes.Buffer) error {
 	timer := prometheus.NewTimer(b.metrics.buildTime)
 	defer timer.ObserveDuration()
 
-	// We reset after a successful flush, but we also reset the buffer before
-	// building for safety.
-	b.flushBuffer.Reset()
+	// Reset the buffer before building for safety.
+	output.Reset()
```

Comment:
IMO, I think resetting the buffer should be the caller's job; it's entirely possible the caller may intend to append a data object following some other data they already appended.